### ISSUE TITLE
[FEA] Aggregation APIs shouldn't take aggregation parameters as unique_ptr

### ISSUE THREAD

**Issue by jrhemstad** [rapidsai/cudf#10393]
**Is your feature request related to a problem? Please describe.**

Some aggregation APIs like `cudf::reduce` take the aggregation parameter as a [`unique_ptr<aggregation> const&`](https://github.com/rapidsai/cudf/blob/1e5b01f1f1e4e97b930de45074dc7109e22a9744/cpp/include/cudf/reduction.hpp#L68).

They should not do this. There is effectively[ _never_ a reason to pass a parameter as `unique_ptr<T> const&` ](https://www.modernescpp.com/index.php/c-core-guidelines-passing-smart-pointer).

**Describe the solution you'd like**

Pass as a `reduce_aggregation const&` instead.

**Comment by vyasr** [comment]
@jrhemstad I agree with this in principle, but I wonder if this might make a worse user experience in practice. Does anyone actually create aggregations in any way aside from using the `make_*_aggregation` factories? Those all return `std::unique_ptr<aggregation>`. Yes, the current API for reductions makes ownership assumptions about the inputs that make it less flexible than it should be and enforces that ownership on the caller, but it seems like the tradeoff in this instance would be forcing every single call site to add the dereference `cudf::reduce(..., *cudf::make_*_aggregation<aggregation>())`.

**Comment by jrhemstad** [comment]
> Does anyone actually create aggregations in any way aside from using the make_*_aggregation factories?

It shouldn't be possible to construct via anything other than a factory. But that's not the point.

I remember there was some specific problem we ran into that inspired me to file this issue, but I can't remember exactly. 

You'll never be able to convince me that a function that takes `unique_ptr<T> const&` is anything other than bad and should be changed. Rules like this exist for good reason and invariably you end up wishing you'd just followed them. 

A single character change to add a dereference is vastly preferable to the status quo.

**Comment by vyasr** [comment]
> It shouldn't be possible to construct via anything other than a factory.

At the risk of diverting this conversation, is that a general stance or are you specifically referring to how aggregations work? If the former... why?

> You'll never be able to convince me that a function that takes unique_ptr<T> const& is anything other than bad and should be changed. Rules like this exist for good reason and invariably you end up wishing you'd just followed them.

My understanding of the rule is that it serves two purposes:
1. Communication of intent. If the function does not need to know about the ownership characteristics, it shouldn't act like it needs to.
2. Flexibility. Accepting a reference or a raw pointer argument instead allows you to pass in an object regardless of how it was allocated, whereas the current approach requires a `unique_ptr`.

I agree with both of those in principle, and I'm happy to make the change, but felt that it was worth noting the cost to achieving that level of purity here. If you should never construct these with anything other than a factory, then point 2 above is moot, and only point 1 is relevant.

**Comment by vyasr** [comment]
@jrhemstad I've pushed the fixes, but for my edification I am still interested to hear your answer 🙂

**Comment by jrhemstad** [comment]
>  are you specifically referring to how aggregations work?

Yes. This is just unique how we designed cudf's aggregations. The base `aggregation` type [should be opaque ](https://github.com/rapidsai/cudf/issues/6895) and the only way to construct them should be through the factories. 

That isn't to say the derived aggregation types couldn't be directly constructible. We just chose to make the factories the canonical method since our primary users are runtime interpreted languages.
